File: C++/064. Minimum Path Sum.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int minPathSum(std::vector<std::vector<int>>& grid) {
        if (grid.size() == 0 || grid[0].size() == 0)
            return 0;
        auto m = int(grid.size()), n = int(grid[0].size());
        std::vector<std::vector<int>> dp (m, std::vector<int>(n, 0));
        dp[0][0] = grid[0][0];
        for (int j = 1; j < n; j++)
            dp[0][j] = dp[0][j - 1] + grid[0][j];
        for (int i = 1; i < m; i++)
            dp[i][0] = dp[i - 1][0] + grid[i][0];
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                int Min = dp[i - 1][j] <= dp[i][j - 1] ? dp[i - 1][j] : dp[i][j - 1];
                dp[i][j] = Min + grid[i][j];
            }
        }
        return dp[m - 1][n - 1];
    }
};

File: Go/064. Minimum Path Sum.go
package main

func minPathSum(grid [][]int) int {
	if len(grid) == 0 || len(grid[0]) == 0 {
		return 0
	}
	m, n := len(grid), len(grid[0])
	dp := make([][]int, m, m)
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n, n)
	}
	dp[0][0] = grid[0][0]
	for j := 1; j < n; j++ {
		dp[0][j] = grid[0][j] + dp[0][j - 1]
	}
	for i := 1; i < m; i++ {
		dp[i][0] = grid[i][0] + dp[i - 1][0]
	}
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[i][j] = grid[i][j] + dp[i - 1][j]
			if dp[i][j] > grid[i][j] + dp[i][j - 1] {
				dp[i][j] = grid[i][j] + dp[i][j - 1]
			}
		}
	}
	return dp[m - 1][n - 1]
}